Unify Enum Format - 1

mitchute opened the pull request:
Pull request overview
---------------------
 - First step in unifying `enum class` format.

```c++
enum class MyEnum {Invalid = -1, ValidChoice1, ... ValidChoiceN, Num};
```

There are a few exceptions that eventually need to be removed, and there are still a bunch of unhandled reviewer comments and enums with explicit values. Those will be addressed in the next PR.

mitchute committed:
make enum class 'unassigned' values equal '-1'

Myoldmopar reviewed (approved):
This seems totally harmless, but continuing to push the effort to make our enums consistent.  CI is appearing clean, which indicates that indeed the actual "numbers" used for these enums were not important in the code, as expected.  If CI continues to be green, this should drop right in easily.

amirroth commented on src/EnergyPlus/ChillerReformulatedEIR.hh:
```diff
@@ -67,7 +67,7 @@ namespace ChillerReformulatedEIR {
     enum class PLR
     {
-        Unassigned,
+        Unassigned = -1,
```
A few more notes as you are doing this pass. First, add `Num` at the end.

mitchute replied:
To that point: assuming they also don't have `Unassigned` or `Invalid`, I should add those here?

amirroth replied:
Yes. And that should be the only negative `enum` value. I know there are a few places where we have multiple negative `enum` values, those need to be flagged.

amirroth commented on src/EnergyPlus/CondenserLoopTowers.hh:
```diff
@@ -71,7 +71,7 @@ namespace CondenserLoopTowers {
         // Empirical Model Type
-        Unassigned,
+        Unassigned = -1,
         CoolToolsXFModel,
```
Given that `enum class` is a name scope, avoid including the name of the type in the name of any of its values, e.g., `Model` here. Otherwise, the resolved name is `ModelType::CoolToolsXFModel`.

amirroth commented on src/EnergyPlus/CondenserLoopTowers.hh:
```diff
@@ -92,21 +92,21 @@ namespace CondenserLoopTowers {
     enum class CapacityCtrlEnum
     {
-        Unassigned,
+        Unassigned = -1,
```
I don't really care if we call `enum class` types `X` or `XType` or `XEnum` or `X_T` or `X_E` (`XTypeEnum` is bad because an `enum class` *is* a type), but we should do it consistently.

amirroth commented on src/EnergyPlus/CurveManager.hh:
```diff
@@ -82,7 +82,7 @@ namespace CurveManager {
     {
-        Unassigned,
+        Unassigned = -1,
         Linear,
```
Same comment as above. An `enum` is a type. Calling something `XTypeEnum` is the same as calling it `XTypeType`.

amirroth commented on src/EnergyPlus/DataGenerators.hh:
```diff
@@ -97,21 +97,21 @@ namespace DataGenerators {
     {
-        Unassigned,
+        Unassigned = -1,
         WaterInReformMains,
```
`WaterInReform` (whatever that means) should probably be part of the type name and removed from names of individual type values.

amirroth commented on src/EnergyPlus/DataGenerators.hh:
```diff
@@ -136,22 +136,22 @@ namespace DataGenerators {
     {
-        Unassigned,
+        Unassigned = -1,
         OpModeOff,     // CHP operating mode OFF
```
`OpMode` is not necessary here.

amirroth commented on src/EnergyPlus/DataGenerators.hh:
```diff
@@ -161,7 +161,7 @@ namespace DataGenerators {
     {
-        Unassigned,
+        Unassigned = -1,
         fuelModeGaseousConstituents,
```
Same comment.

amirroth commented on src/EnergyPlus/DataRoomAirModel.hh:
```diff
@@ -109,7 +109,7 @@ namespace DataRoomAirModel {
     {
-        Unassigned,
+        Unassigned = -1,
         OutdoorDryBulbMode,  // by outdoor air bulb.
```
`Mode`.

amirroth commented on src/EnergyPlus/DataRoomAirModel.hh:
```diff
@@ -120,7 +120,7 @@ namespace DataRoomAirModel {
     {
-        Unassigned,
+        Unassigned = -1,
         ConstGradTempPattern,  // constant gradient in vertical direction
```
`Pattern`.

amirroth commented on src/EnergyPlus/DemandManager.hh:
```diff
@@ -69,7 +69,7 @@ namespace DemandManager {
     {
-        Unassigned,
+        Unassigned = -1,
         ManagerTypeExtLights,
```
`ManagerType`.

amirroth commented on src/EnergyPlus/DemandManager.hh:
```diff
@@ -79,15 +79,15 @@ namespace DemandManager {
     enum class ManagePriorityType
     {
-        Unassigned,
+        Unassigned = -1,
```
I can keep making the same comment, but you get the idea.

mitchute committed:
Merge branch 'develop' into enum-null-state

mitchute committed:
add 'Invalid' and 'Num' to enum classes, partially commplete

mitchute committed:
Merge branch 'develop' into enum-null-state

mitchute committed:
fix a few constexprs on wrapped lines

mitchute committed:
add script to check for malformed enums

mitchute committed:
fix test failures, more enum cleanups

mitchute committed:
rollback EPWWeather.f90 changes

mitchute committed:
enums cleanup

mitchute committed:
move 'State' to 'state' for EnergyPlusData

mitchute committed:
fix build errors/warnings

mitchute committed:
Merge branch 'develop' into enum-null-state

mitchute committed:
fix diffs

mitchute committed:
'unassigned' to 'invalid', 'count' to 'num', unify value casing for 'invalid' and 'num'

mitchute committed:
first round of making enum names PascalCase

mitchute committed:
fix '-fpermissive' build error caused by identically named enum and var

mitchute committed:
second round of making enum names PascalCase, some renaming for clarity

mitchute committed:
fix -fpermissive error

mitchute committed:
another round of making enum names PascalCase

mitchute committed:
another round of making enum names PascalCase

mitchute committed:
another round of making enum names PascalCase

mitchute committed:
another round of making enum names PascalCase

mitchute committed:
Merge branch 'develop' into enum-null-state

mitchute committed:
fix build error, cleanup

mitchute committed:
more cleanup, update enum check script

mitchute committed:
make enum key names PascalCase

mitchute committed:
rename enum keys that are too similar to enum name

mitchute committed:
fix some enum names, fix eio diffs

mitchute committed:
a few final cleanups, turn off some check for now, and add a few exceptions

mitchute commented:
@Myoldmopar @amirroth I'll be watching the results for this latest commit, but I'm hoping this PR is nearly done for now. I'll punt addressing all of the comments to the next one. There are still enums that have explicit values but in going down the road of eliminating them, I ran into problems that are going to cause diffs. So I'd rather hold anything like that until it can be done standalone with a much smaller set of changes.

amirroth commented:
> @Myoldmopar @amirroth I'll be watching the results for this latest commit, but I'm hoping this PR is nearly done for now. I'll punt addressing all of the comments to the next one. There are still enums that have explicit values but in going down the road of eliminating them, I ran into problems that are going to cause diffs. So I'd rather hold anything like that until it can be done standalone with a much smaller set of changes.

Sounds fine. I appreciate you writing a script to root out all of these things, will make it easier to keep enforcing this pattern.

mitchute committed:
fix diffs

mitchute commented:
It's looking like this is going to come back clean at this point. Unless something else pops up, I'll pass it over to @Myoldmopar for review/merge.

mitchute committed:
Merge branch 'develop' into enum-null-state

mjwitte commented:
@mitchute Curious - what's in here that would cause the performance measure to be -0.32%?

mitchute commented:
Nothing that I'm aware of. It's all just enum renaming and adding `Invalid` and `Num` where missing. Is 0.32% still at the "noise" level, or is that a meaningful number?

Myoldmopar commented:
Yeah I was poking around and didn't see anything, but it's hard to identify anything amidst the size of these changes.  And I can't imagine what about this would cause a performance degradation.  0.32% generally isn't in the noise, but in this case I'm not sure.

amirroth commented:
The only thing I can think of is that the number for some of these enums has changed (a few were changed from starting with 1 to starting with 0), but I don't see how that would make a difference either. If anything, it should make things faster. 

I appreciate the instinct to look into this, that's the right mindset. But I wouldn't spend a ton of time on this and I suspect it's going to be hard to find out what's going on in a reasonable amount of time. Hence, probably okay to move on.

mjwitte commented:
Actually, most of the files in the performance tests improve slightly, performance.PipingSystem_Underground_FHX negates all of that and then some. Looks like this function may be the source of the increase `EnergyPlus::PlantPipingSystemsManager::Domain::EvaluateFieldCellTemperature(EnergyPlus::PlantPipingSystemsManager::CartesianCell&` and `PerformPipeCellSimulation`? Below that the numbers are identical.

mitchute commented:
Is there a clean way to see the side-by-side results for these? I might have to rig up a web scraping script in the future if not. 

Perhaps `PerformTemperatureFieldUpdate` is causing this? https://github.com/NREL/EnergyPlus/blob/a4789f2517667deef6c5c61e271e0e2f1bac90d7/src/EnergyPlus/PlantPipingSystemsManager.cc#L4147

I added a `default: assert(false)` to that `switch/case` statement for the `Invalid` and `Num` values.

amirroth commented:
> I appreciate the instinct to look into this, that's the right mindset. But I wouldn't spend a ton of time on this and I suspect it's going to be hard to find out what's going on in a reasonable amount of time. Hence, probably okay to move on.

Scratch that, maybe it's not so hard to find.

amirroth commented:
All I will say is that it figures that this would happen in a routine written by @Myoldmopar. That guy!

mitchute commented:
Is `default: assert(false)` not the right way to handle these? Or should all `switch/case` blocks just be exhaustive from here on out?

mjwitte commented:
> Is there a clean way to see the side-by-side results for these? I might have to rig up a web scraping script in the future if not.

No, there isn't a clean way and because develop's results move, the performance measure display can quickly get out of date. It would be really nice for the CI results to produce a static table of each file with develop, this commit, delta, and delta %, maybe? More detail below that would be useful, but that could get messy to sort out function names if they don't align.



> Is `default: assert(false)` not the right way to handle these? Or should all `switch/case` blocks just be exhaustive from here on out?

That line should never execute, esp with a non-debug build, right? Is it possible there's someplace that's repeatedly instantiating one or more of these enums as an array so that the extra members are adding instructions? Just fishing here.

mjwitte commented:
Or does a `default` case add some instructions somehow. It's interesting to see where this is possibly coming from, and it's from a less commonly used feature, so it's not critical.

mitchute commented:
IDK. I always have to look slightly askance at the performance results. My inclination is that this particular change should *not* make a meaningful difference. Happy to make any changes though, as recommended by yon "seasoned" developers. 🧙‍♂️

amirroth commented:
> Or does a `default` case add some instructions somehow. It's interesting to see where this is possibly coming from, and it's from a less commonly used feature, so it's not critical.

No, that's the way to do it. It doesn't add instructions if the `default` case is not actually hit. But at that point the simulation is over so who cares.

amirroth commented:
I think it's time to move on.

mjwitte commented:
It's just this isn't the first one of these sweeping PRs that's bumped up the performance count slightly for no apparent reason. Oh well, what-ever, it's fine, really, just fine, fine fine fine, said the well-seasoned one.

amirroth commented:
> It's just this isn't the first one of these sweeping PRs that's bumped up the performance count slightly for no apparent reason. Oh well, what-ever, it's fine, really, just fine, fine fine fine, said the well-seasoned one.

Okay. I'm not sure that this means that we should hold up this PR, but it does sound like we need better CI support for tracking and flagging these kinds of things. @Myoldmopar.

Myoldmopar commented:
I'm caught up with everything here and going to merge this in.  Thanks everyone!
